File: src/Ogooreck.Sample.BusinessLogic.Tests/Deciders/BankAccountDecider.cs
namespace Ogooreck.Sample.BusinessLogic.Deciders;

public record OpenBankAccount(
    Guid BankAccountId,
    string AccountNumber,
    Guid ClientId,
    string CurrencyISOCode
);

public record RecordDeposit(
    decimal Amount,
    Guid CashierId
);

public record WithdrawnCashFromATM(
    decimal Amount,
    Guid AtmId
);

public record CloseBankAccount(
    string Reason
);

public static class BankAccountDecider
{

    public static object Handle(
        Func<DateTimeOffset> now,
        object command,
        BankAccount bankAccount
        ) =>
        command switch
        {
            OpenBankAccount openBankAccount =>
                Handle(now, openBankAccount),
            RecordDeposit recordDeposit =>
                Handle(now, recordDeposit, bankAccount),
            WithdrawnCashFromATM withdrawnCash =>
                Handle(now, withdrawnCash, bankAccount),
            CloseBankAccount closeBankAccount =>
                Handle(now, closeBankAccount, bankAccount),
            _ =>
                throw new InvalidOperationException($"{command.GetType().Name} cannot be handled for Bank Account")
        };

    private static BankAccountOpened Handle(
        Func<DateTimeOffset> now,
        OpenBankAccount command
    ) =>
        new BankAccountOpened(
            command.BankAccountId,
            command.AccountNumber,
            command.ClientId,
            command.CurrencyISOCode,
            now(),
            1
        );

    private static DepositRecorded Handle(
        Func<DateTimeOffset> now,
        RecordDeposit command,
        BankAccount account
    )
    {
        if (account.Status == BankAccountStatus.Closed)
            throw new InvalidOperationException("Account is closed!");

        return new DepositRecorded(account.Id, command.Amount, command.CashierId, now(), account.Version + 1);
    }

    private static CashWithdrawnFromATM Handle(
        Func<DateTimeOffset> now,
        WithdrawnCashFromATM command,
        BankAccount account
    )
    {
        if (account.Status == BankAccountStatus.Closed)
            throw new InvalidOperationException("Account is closed!");

        if (account.Balance < command.Amount)
            throw new InvalidOperationException("Not enough money!");

        return new CashWithdrawnFromATM(account.Id, command.Amount, command.AtmId, now(), account.Version + 1);
    }

    private static  BankAccountClosed Handle(
        Func<DateTimeOffset> now,
        CloseBankAccount command,
        BankAccount account
    )
    {
        if (account.Status == BankAccountStatus.Closed)
            throw new InvalidOperationException("Account is already closed!");

        return new BankAccountClosed(account.Id, command.Reason, now(), account.Version + 1);
    }
}

File: src/Ogooreck.Sample.BusinessLogic.FSharp.Tests/Deciders/BankAccountDecider.fs
module Deciders.BankAccountDecider

open System
open BankAccount
open BankAccountPrimitives

type OpenBankAccount =
    { BankAccountId: AccountId
      AccountNumber: AccountNumber
      ClientId: ClientId
      CurrencyIsoCode: CurrencyIsoCode
      Now: DateTimeOffset }

type RecordDeposit =
    { Amount: decimal
      CashierId: CashierId
      Now: DateTimeOffset }

type WithdrawCashFromAtm =
    { Amount: decimal
      AtmId: AtmId
      Now: DateTimeOffset }

type CloseBankAccount = { Reason: string; Now: DateTimeOffset }

type Command =
    | OpenBankAccount of OpenBankAccount
    | RecordDeposit of RecordDeposit
    | WithdrawCashFromAtm of WithdrawCashFromAtm
    | CloseBankAccount of CloseBankAccount

let openBankAccount (command: OpenBankAccount) bankAccount : Event =
    match bankAccount with
    | Open _ -> invalidOp "Account is already opened!"
    | Closed _ -> invalidOp "Account is already closed!"
    | Initial ->
        BankAccountOpened
            { BankAccountId = command.BankAccountId
              AccountNumber = command.AccountNumber
              ClientId = command.ClientId
              CurrencyIsoCode = command.CurrencyIsoCode
              CreatedAt = command.Now
              Version = 1 }

let recordDeposit (command: RecordDeposit) bankAccount : Event =
    match bankAccount with
    | Initial -> invalidOp "Account is not opened!"
    | Closed _ -> invalidOp "Account is closed!"
    | Open state ->
        DepositRecorded
            { BankAccountId = state.Id
              Amount = command.Amount
              CashierId = command.CashierId
              RecordedAt = command.Now
              Version = state.Version + 1L }

let withdrawCashFromAtm (command: WithdrawCashFromAtm) bankAccount : Event =
    match bankAccount with
    | Initial -> invalidOp "Account is not opened!"
    | Closed _ -> invalidOp "Account is closed!"
    | Open state ->
        if (state.Balance < command.Amount) then
            invalidOp "Not enough money!"

        CashWithdrawnFromAtm
            { BankAccountId = state.Id
              Amount = command.Amount
              AtmId = command.AtmId
              RecordedAt = command.Now
              Version = state.Version + 1L }

let closeBankAccount (command: CloseBankAccount) bankAccount : Event =
    match bankAccount with
    | Initial -> invalidOp "Account is not opened!"
    | Closed _ -> invalidOp "Account is already closed!"
    | Open state ->
        BankAccountClosed
            { BankAccountId = state.Id
              Reason = command.Reason
              ClosedAt = command.Now
              Version = state.Version + 1L }

let decide command bankAccount =
    match command with
    | OpenBankAccount c -> openBankAccount c bankAccount
    | RecordDeposit c -> recordDeposit c bankAccount
    | WithdrawCashFromAtm c -> withdrawCashFromAtm c bankAccount
    | CloseBankAccount c -> closeBankAccount c bankAccount
